Robust type name to type identifier conversion for C harnesses

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
This PR makes the type name to type identifier conversion used in generating C harnesses more robust.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] n/a Because this improves existing functionality, to reduce potential compile errors in generated harnesses and I assume that these potential errors are not documented. ~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] n/a - Non claimed ~My commit message includes data points confirming performance improvements (if claimed).~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Document `type2identifier` function

In order to specify the properties which are expected to hold for this
function.

thomasspriggs committed:
Refactor `type2identifier`, splitting into 2 functions

In order to make the conversion from type name to type identifier unit
testable in isolation.

thomasspriggs committed:
Add unit test of existing `type_name2type_identifier` behaviour

In order to quickly test that no regressions have occured as part of
adding extra functionality.

thomasspriggs committed:
Add more robust type name to type identifier conversion

In order to certain that we don't emmit C harnesses containing invalid
identifiers.

The unit test includes both a function to generate the string of all
printable characters and the same string as a string literal in order to
show that the string does include all printable characters and to show
what these characters are.

NlightNFotis commented on src/ansi-c/type2name.cpp:
```diff
@@ -301,3 +300,8 @@ std::string type2identifier(const typet &type, const namespacet &ns)
+
+std::string type2identifier(const typet &type, const namespacet &ns)
+{
+  return type_name2type_identifier(type2name(type, ns));
```
Nice :+1:

NlightNFotis reviewed (approved):
Solid!

hannes-steffenhagen-diffblue commented on src/ansi-c/type2name.cpp:
```diff
@@ -277,27 +279,52 @@ std::string type2name(const typet &type, const namespacet &ns)
 /// should ideally always be valid C identifiers
 /// This replaces some invalid characters that can appear in type2name output.
-std::string type2identifier(const typet &type, const namespacet &ns)
+std::string type_name2type_identifier(const std::string &name)
```
I still don’t like the `2` naming convention here, but changing that isn’t really in scope for this PR.

hannes-steffenhagen-diffblue commented on src/ansi-c/type2name.cpp:
```diff
@@ -277,27 +279,52 @@ std::string type2name(const typet &type, const namespacet &ns)
+    };
+  const auto remove_duplicate_underscores = [](const std::string &identifier) {
+    static const std::regex duplicate_underscore{"_+"};
+    return std::regex_replace(identifier, duplicate_underscore, "_");
```
I’m not completely sure this is always ‘the right thing’ to do, for example some compilers (including `gcc`) have a builtin `__int128` type, i.e. double underscores in names are a relatively normal thing to see.

That being said, since we only really need these to be related to the original names and not _necessarily_ completely match up with them I suppose this is fine.

chrisr-diffblue replied:
Yeah, I'm not sure this is the right thing to do either - double underscore is a perfectly valid identifier for internal symbols - which you may well get in things like header files.

thomasspriggs replied:
Superflous double underscores could be generated in the preceding section of code. For example `char**` would be replaced with `char_ptr__ptr`, which is what I wanted to tidy up.

However I would also argue that we should not emit identifiers which begin with an underscore in the harness code. This is because identifiers which start with an underscore are reserved for use by the compiler / standard libraries. See section 7.1.3 of the standard - http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf
I would argue that because we are outputting the  harnesses as C code for user modification, the harnesses are therefore user code. This means that use of identifiers with leading underscores would invoke "undefined behaviour" in terms of the language standards.

hannes-steffenhagen-diffblue replied:
@thomasspriggs note that this is currently not necessarily being used only for the harness – that’s where we found the issue, but the reason it occurred was actually because of the way we generated names for specialisations of generic builtin functions.

chrisr-diffblue replied:
I know that double-underscore is reserved for compiler/library use, and partly thats why I think it may be ok to emit them - because if you are generating a C file from a goto-binary, that goto-binary may well have legitimate typenames from include files (e.g. from things like #defines, etc) and as such, they will be at least semi-legitimate to emit (assuming the emitted code is then compiled against the same headers/libraries). 

I may of course be missing something about the context for this change, so if you have an example of where this fails and you need this, do feel free to point me at it :-)

thomasspriggs replied:
Ok. I have updated this, in order to retain more underscores.

chrisr-diffblue commented on src/ansi-c/type2name.cpp:
```diff
@@ -277,27 +279,52 @@ std::string type2name(const typet &type, const namespacet &ns)
+  const auto replace_invalid_characters_with_underscore =
+    [](const std::string &identifier) {
+      static const std::regex non_alpha_numeric{"[^A-Za-z0-9]"};
+      return std::regex_replace(identifier, non_alpha_numeric, "_");
```
This, I assume, risks generating clashing names?

hannes-steffenhagen-diffblue replied:
@chrisr-diffblue Yes, in general this isn’t generating a unique identifier. The idea would be to use this as an identifier fragment useful for generating a human-readable-but-unique identifier elsewhere.

thomasspriggs replied:
We were already risking clashing names, rather than guaranteeing uniqueness. For example `fruit*` will be replaced with `fruit_ptr_`, but there was no guarantee that something else in user code wasn't already called `fruit_ptr`.

chrisr-diffblue commented on src/ansi-c/type2name.h:
```diff
@@ -20,6 +20,15 @@ class namespacet;
+/**
+ * Constructs a string describing the given type, which can be used as part of
+ * a `C` identifier. The resulting identifier is not guaranteed to uniquely
+ * identify the type in all cases.
```
This might need clarifying  - in fact, it might actually be worth making this explicit in the function name, e.g. rename it to something like `type_to_approximate_identifier` or `type_to_nonunique_identifier`.

thomasspriggs replied:
I get that "identifier is a bit of a misnomer in this case. However `nonunique_identifier` seems like a bit of an oxymoron to me. How does the term `partial_identifier` sound to you?

chrisr-diffblue replied:
`partial_identifer` sounds reasonable to me - it makes it clearer to an unsuspecting caller that they need to understand what is returning :-)

thomasspriggs replied:
Done.

thomasspriggs committed:
Rename `type2identifier` to `type_to_partial_identifier`

Because the return value of this function does not uniquely identify a
type, it is only used as part of a unique identifier.

thomasspriggs committed:
Remove duplicate underscore removal

In order to retain duplicate underscores from original type identifiers.

Invalid character removal is also tweaked to reduce the number of
underscores introduced in the previous step.

thomasspriggs committed:
Make `type_name2type_identifier` retain UTF-8 characters

In order to better support UTF-8.

thomasspriggs committed:
Make `type_name2type_identifier` retain leading underscores

Because the resulting partial identifiers are considered to be internal,
it is fine for them to start with an underscore.

thomasspriggs commented:
@chrisr-diffblue Please can you re-review? Comments are addressed in new commits at the end of the PR.

thomasspriggs commented:
I have pushed to restart CI only. The windows job appeared to fail for reasons unrelated to my changes.

chrisr-diffblue reviewed (approved):
Looks good, thanks. Approving, with a slight note of caution that if CBMC ever supports (or claims to support) internationalised source code (i.e. getting into the whole input file encoding/internal encoding question) then this might need revisiting. It won't be the only place that'll need revisiting though :-)

thomasspriggs commented:
I did take a look at the implementation of the lexer and there does seem to be some UTF-8 support in there. But it is reasonable to assume that there will be incompatibilities, if there are no end-to-end regression tests for such functionality.
